File: step05listview/src/main/java/com/example/step05listview/MainActivity.java
package com.example.step05listview;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements AdapterView.OnItemClickListener, AdapterView.OnItemLongClickListener {

    //필드
    List<String> names;
    ArrayAdapter<String> adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //ListView의 참조값
        ListView listView = findViewById(R.id.listView);

        //list에 호출할 sample data
        List<String> names = new ArrayList<>();
        names.add("김구라");
        names.add("해골");
        names.add("원숭이");
        for(int i=0; i<10; i++){
            names.add("아무개"+i);
        }

        //simple_list_item_1 : list view에 올릴 model들(여기선 List<String>)의 하나하나의 layout을 담당한다.
        adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1,
                names);

        //ListView에 adapter연결하기
        listView.setAdapter(adapter);
        
        //각각의 아이템을 클릭했는지 리슨하는 객체
        listView.setOnItemClickListener(this);
        listView.setOnItemLongClickListener(this);
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        // int i는 클릭한 아이템의 index가 들어있다.
        String name = names.get(i);
        Toast.makeText(this,name,Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
        //오랫동안 클릭한 셀에 출력된 이름 읽어오기
        String name = names.get(i);
        //아래의 익명 클래스에서 참조가 가능하도록
        int selectedIndex = i;
        
        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage(name+"을 삭제하시겠습니까?")
                .setNegativeButton("아니요", null)
                .setPositiveButton("네", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int i) {
                        //모델에서 해당 인덱스의 data를 삭제한다.
                        names.remove(selectedIndex);
                        //adapter에 모델에 변화가 생겼다고 알려준다.
                        adapter.notifyDataSetChanged();
                    }
                })
                .create()
                .show();
        //이벤트 전파를 여기서 막기

        return false;
    }
}

File: step14sqlite/src/main/java/com/example/step14sqlite/MainActivity.kt
package com.example.step14sqlite

import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.view.View.OnClickListener
import android.widget.*
import android.widget.AdapterView.OnItemLongClickListener
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity(), OnClickListener, OnItemLongClickListener {
    
    //필요한 필드 정의하기

    //null을 대입했다가 나중에 값을 바꾸려면 번거롭다.
    lateinit var inputText: EditText //lateinit예약어를 활용하면 null을 넣을 필요 없이 값을 나중에 얻을 수 있다.
    lateinit var listView:ListView
    lateinit var adapter:TodoAdapter
    lateinit var helper:DBHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //필요한 UI의 참조값을 얻어와서 필드에 저장하기
        inputText = findViewById(R.id.inputText)
        listView = findViewById(R.id.listView)

        //버튼 리스너 등록하기
        findViewById<Button>(R.id.addBtn).setOnClickListener(this)
        
        //DBHelper객체의 참조값을 필드에 저장하기
        helper = DBHelper(this, "MyDB.sqlite", null, 1)

        //할 일 목록 얻어오기  // type은 추론이 가능하므로 List<Todo>는 생략 가능
        val list:List<Todo> = TodoDao(helper).getList()

        //listView동작 준비하고, 할 일 목록 출력하기
        adapter = TodoAdapter(this, R.layout.listview_cell, list )
        listView.adapter = adapter

        listView.setOnItemLongClickListener(this)
    }


    override fun onClick(v: View?) {
        //1. 입력한 문자열을 읽어와서
        val msg = inputText.text.toString()
        //2. Todo객체에 담아서
        val data = Todo(0, msg, "")//빈 생성자를 억지로 만들어줌
        //3. TodoDao객체를 이용해서 DB에 저장한다.
        TodoDao(helper).insert(data)
        //4. 목록을 새로 얻어와서
        val list = TodoDao(helper).getList()
        //5. adapter에 넣어주고
        adapter.list = list
        //6. model의 내용이 바뀌었다고 adapter에 달려서 ListView가 업데이트 되도록 한다.
        adapter.notifyDataSetChanged()
        //7. Toast띄우기
        Toast.makeText(this, "저장했습니다", Toast.LENGTH_SHORT).show()
        inputText.setText("")

        //8. ListView의 가장 아래쪽이 화면에 보일 수 있도록 부드럽게 스크롤 시키기
        listView.smoothScrollToPosition(adapter.count)
    }

    override fun onItemLongClick(
        parent: AdapterView<*>?,
        view: View?,
        position: Int,
        id: Long
    ): Boolean {
        // 여기서 position 은 클릭한 cell 의 인덱스
        // 여기서 id 는 클릭한 cell 의 아이디 ( Todo 의 primary key )
        // id 에 전달되는 값은 TodoAdapter 의  getItemId() 메소드에서 리턴한 값

        AlertDialog.Builder(this)
            .setTitle("알림")
            .setMessage("삭제 하시겠습니까?")
            .setPositiveButton("네", DialogInterface.OnClickListener { dialog, which ->
                val dao=TodoDao(helper)
                dao.delete(id.toInt())
                //목록을 새로 얻어와서
                val list=TodoDao(helper).getList()
                //아답타에 넣어주고
                adapter.list=list
                //모델의 내용이 바뀌었다고 아답타에 알려서 ListView 가 업데이트 되도록 한다.
                adapter.notifyDataSetChanged()
            })
            .setNegativeButton("아니요", null)
            .create()
            .show()

        return false
    }
}